ucp/_libs/ucp_py.pyx

```diff
@@ -8,6 +8,7 @@ import sys
 import logging
 import os
 import socket
+from pyroute2 import IPRoute
```

Comment:
I'm not sure if this is worth adding a 3rd party dependency. Perhaps something similar can be done using the socket module (which is part of the standard library).

Reply:
I wasn't too lucky with the standard libraries with this finiding interfaces/routes for a given IP (where listener is listening) but I'm happy to switch if there's a suggestion for this.

Reply:
So you want to go from an IP address like `127.0.0.1` to a network interface like `lo` ?

Reply:
It's not standard-lib, but the `psutil` module might have you covered here, and it's much more common/trusted than `pyroute2`.

```python
In [1]: import psutil

In [2]: psutil.net_if_addrs()
Out[2]:
{'lo': [snicaddr(family=<AddressFamily.AF_INET: 2>, address='127.0.0.1', netmask='255.0.0.0', broadcast=None, ptp=None),
  snicaddr(family=<AddressFamily.AF_INET6: 10>, address='::1', netmask='ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff', broadcast=None, ptp=None),
  snicaddr(family=<AddressFamily.AF_PACKET: 17>, address='00:00:00:00:00:00', netmask=None, broadcast=None, ptp=None)],
 'enp1s0f0': [snicaddr(family=<AddressFamily.AF_INET: 2>, address='10.33.227.165', netmask='255.255.255.0', broadcast='10.33.227.255', ptp=None),
  snicaddr(family=<AddressFamily.AF_INET6: 10>, address='fe80::dac4:97ff:fe64:f624%enp1s0f0', netmask='ffff:ffff:ffff:ffff::', broadcast=None, ptp=None),
  snicaddr(family=<AddressFamily.AF_PACKET: 17>, address='d8:c4:97:64:f6:24', netmask=None, broadcast='ff:ff:ff:ff:ff:ff', ptp=None)],
 'ib0': [snicaddr(family=<AddressFamily.AF_INET: 2>, address='10.33.225.165', netmask='255.255.254.0', broadcast='10.33.225.255', ptp=None),
  snicaddr(family=<AddressFamily.AF_INET6: 10>, address='fe80::526b:4b03:28:d52e%ib0', netmask='ffff:ffff:ffff:ffff::', broadcast=None, ptp=None),
  snicaddr(family=<AddressFamily.AF_PACKET: 17>, address='80:00:00:68:fe:80:00:00:00:00:00:00:50:6b:4b:03:00:28:d5:2e', netmask=None, broadcast='00:ff:ff:ff:ff:12:40:1b:ff:ff:00:00:00:00:00:00:ff:ff:ff:ff', ptp=None)],
 'ib1': [snicaddr(family=<AddressFamily.AF_INET: 2>, address='10.33.225.175', netmask='255.255.254.0', broadcast='10.33.225.255', ptp=None),
  snicaddr(family=<AddressFamily.AF_INET6: 10>, address='fe80::526b:4b03:35:e806%ib1', netmask='ffff:ffff:ffff:ffff::', broadcast=None, ptp=None),
  snicaddr(family=<AddressFamily.AF_PACKET: 17>, address='80:00:00:68:fe:80:00:00:00:00:00:00:50:6b:4b:03:00:35:e8:06', netmask=None, broadcast='00:ff:ff:ff:ff:12:40:1b:ff:ff:00:00:00:00:00:00:ff:ff:ff:ff', ptp=None)],
 'ib2': [snicaddr(family=<AddressFamily.AF_INET: 2>, address='10.33.225.185', netmask='255.255.254.0', broadcast='10.33.225.255', ptp=None),
  snicaddr(family=<AddressFamily.AF_INET6: 10>, address='fe80::526b:4b03:1b:6dac%ib2', netmask='ffff:ffff:ffff:ffff::', broadcast=None, ptp=None),
  snicaddr(family=<AddressFamily.AF_PACKET: 17>, address='80:00:00:68:fe:80:00:00:00:00:00:00:50:6b:4b:03:00:1b:6d:ac', netmask=None, broadcast='00:ff:ff:ff:ff:12:40:1b:ff:ff:00:00:00:00:00:00:ff:ff:ff:ff', ptp=None)],
 'ib3': [snicaddr(family=<AddressFamily.AF_INET: 2>, address='10.33.225.195', netmask='255.255.254.0', broadcast='10.33.225.255', ptp=None),
  snicaddr(family=<AddressFamily.AF_INET6: 10>, address='fe80::526b:4b03:1b:6d98%ib3', netmask='ffff:ffff:ffff:ffff::', broadcast=None, ptp=None),
  snicaddr(family=<AddressFamily.AF_PACKET: 17>, address='80:00:00:68:fe:80:00:00:00:00:00:00:50:6b:4b:03:00:1b:6d:98', netmask=None, broadcast='00:ff:ff:ff:ff:12:40:1b:ff:ff:00:00:00:00:00:00:ff:ff:ff:ff', ptp=None)],
 'docker0': [snicaddr(family=<AddressFamily.AF_INET: 2>, address='172.17.0.1', netmask='255.255.0.0', broadcast='172.17.255.255', ptp=None),
  snicaddr(family=<AddressFamily.AF_INET6: 10>, address='fe80::42:90ff:fe92:b546%docker0', netmask='ffff:ffff:ffff:ffff::', broadcast=None, ptp=None),
  snicaddr(family=<AddressFamily.AF_PACKET: 17>, address='02:42:90:92:b5:46', netmask=None, broadcast='ff:ff:ff:ff:ff:ff', ptp=None)],
 'vethb53955c': [snicaddr(family=<AddressFamily.AF_INET6: 10>, address='fe80::18dc:49ff:fed6:fd17%vethb53955c', netmask='ffff:ffff:ffff:ffff::', broadcast=None, ptp=None),
  snicaddr(family=<AddressFamily.AF_PACKET: 17>, address='1a:dc:49:d6:fd:17', netmask=None, broadcast='ff:ff:ff:ff:ff:ff', ptp=None)],
 'enp1s0f1': [snicaddr(family=<AddressFamily.AF_PACKET: 17>, address='d8:c4:97:64:f6:25', netmask=None, broadcast='ff:ff:ff:ff:ff:ff', ptp=None)]}
```

Reply:
> So you want to go from an IP address like 127.0.0.1 to a network interface like lo ?

@mrocklin I meant for discovery outside of local machine. I want to be able to get the IP + interface used to reach a valid IP (192.168.X.Y) from the system. Does `psutil` provide this capability?